qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -266,63 +251,198 @@ def run(self, dag):
+                    cargs,
                 )
-                apply_gate(
-                    mapped_dag,
-                    node,
-                    original_layout,
+            )
+
+        # Note: num_qubits and num_clbits are used to preallocate
+        # storage within SabreDAG. Because indices are in terms of the
+        # root DAG, we always pass the same values even though blocks
+        # will often have fewer bits than the outer circuit.
+        return SabreDAG(num_qubits, num_clbits, dag_list, node_blocks)
+
+    return process_dag(dag, qubit_indices, clbit_indices), circuit_to_dag_dict
+
+
+def _apply_sabre_result(
+    mapped_dag,
+    root_dag,
+    qubit_indices,
+    canonical_register,
+    initial_layout,
+    sabre_result,
+    circuit_to_dag_dict,
+    component_map=None,
+):
+    bit_to_qreg_idx = {bit: idx for idx, bit in enumerate(canonical_register)}
+
+    def empty_dag(node):
+        out = DAGCircuit()
+        for qreg in mapped_dag.qregs.values():
+            out.add_qreg(qreg)
+        out.add_clbits(node.cargs)
+        return out
+
+    def apply_inner(out_dag, current_layout, qubit_indices_inner, result, id_to_node):
+        for node_id in result.node_order:
+            node = id_to_node[node_id]
+            if isinstance(node.op, ControlFlowOp):
+                # Handle control flow op and continue.
+                block_results = result.node_block_results[node_id]
+                mapped_block_dags = []
+                idle_qubits = set(out_dag.qubits)
+                for block, block_result in zip(node.op.blocks, block_results):
+                    block_id_to_node = circuit_to_dag_dict[id(block)]._multi_graph
+                    mapped_block_dag = empty_dag(node)
+                    mapped_block_layout = current_layout.copy()
+                    block_qubit_indices = {
+                        inner: qubit_indices_inner[outer]
+                        for inner, outer in zip(block.qubits, node.qargs)
+                    }
+                    apply_inner(
+                        mapped_block_dag,
+                        mapped_block_layout,
+                        block_qubit_indices,
+                        block_result.result,
+                        block_id_to_node,
+                    )
+
+                    # Apply swap epilogue to bring each block to the same
+                    # final layout.
+                    process_swaps(
+                        block_result.swap_epilogue,
+                        mapped_block_dag,
+                        mapped_block_layout,
+                        canonical_register,
+                        False,
+                        bit_to_qreg_idx,
+                        component_map,
+                    )
+
+                    # If the swap epilogue didn't return us to the initial layout,
+                    # there's a bug.
+                    assert mapped_block_layout.layout_mapping() == current_layout.layout_mapping()
```

Comment:
Is it worth leaving this runtime assertion in place?  It has quite a non-trivial cost to evaluate, especially for large circuits.  We probably want to catch this kind of thing with targetted unit tests, rather than deferring the testing to users' machines.

Reply:
Addressed in 20fe8b4.